src/formats.rs

```diff
@@ -0,0 +1,268 @@
+
+    fn parse_records(lines: I) -> Self::RecordStream;
+}
+
+pub trait RecordWriter {
+    async fn write_record(
+        record: &SequencedRecord,
+        writer: &mut (impl AsyncWrite + Unpin),
+    ) -> io::Result<()>;
+}
+
+pub use text::Formatter as TextFormatter;
+pub type JsonFormatter = json::Formatter<false>;
+pub type JsonBinsafeFormatter = json::Formatter<true>;
+
+mod text {
+    use std::{
+        io,
+        pin::Pin,
+        task::{Context, Poll},
+    };
+
+    use futures::{Stream, StreamExt};
+    use s2::types::{AppendRecord, SequencedRecord};
+    use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+    use super::{RecordParseError, RecordParser, RecordWriter};
+
+    pub struct Formatter;
+
+    impl RecordWriter for Formatter {
+        async fn write_record(
+            record: &SequencedRecord,
+            writer: &mut (impl AsyncWrite + Unpin),
+        ) -> io::Result<()> {
+            let s = String::from_utf8_lossy(&record.body);
+            writer.write_all(s.as_ref().as_bytes()).await
+        }
+    }
+
+    impl<I> RecordParser<I> for Formatter
+    where
+        I: Stream<Item = io::Result<String>> + Send + Unpin,
+    {
+        type RecordStream = RecordStream<I>;
+
+        fn parse_records(lines: I) -> Self::RecordStream {
+            RecordStream(lines)
+        }
+    }
+
+    pub struct RecordStream<S>(S);
+
+    impl<S> Stream for RecordStream<S>
+    where
+        S: Stream<Item = io::Result<String>> + Send + Unpin,
+    {
+        type Item = Result<AppendRecord, RecordParseError>;
+
+        fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            match self.0.poll_next_unpin(cx) {
+                Poll::Pending => Poll::Pending,
+                Poll::Ready(None) => Poll::Ready(None),
+                Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(e.into()))),
+                Poll::Ready(Some(Ok(s))) => Poll::Ready(Some(Ok(AppendRecord::new(s)?))),
+            }
+        }
+    }
+}
+
+mod json {
+    use std::{
+        borrow::Cow,
+        io,
+        pin::Pin,
+        task::{Context, Poll},
+    };
+
+    use base64ct::{Base64, Encoding};
+    use bytes::Bytes;
+    use futures::{Stream, StreamExt};
+    use s2::types::{AppendRecord, AppendRecordParts, ConvertError, Header, SequencedRecord};
+    use serde::{Deserialize, Serialize};
+    use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+    use super::{RecordParseError, RecordParser, RecordWriter};
+
+    #[derive(Debug, Clone, Default)]
+    struct CowStr<'a, const BIN_SAFE: bool>(Cow<'a, str>);
+
+    type OwnedCowStr<const BIN_SAFE: bool> = CowStr<'static, BIN_SAFE>;
+
+    impl<'a, const BIN_SAFE: bool> From<&'a [u8]> for CowStr<'a, BIN_SAFE> {
+        fn from(value: &'a [u8]) -> Self {
+            Self(if BIN_SAFE {
+                Base64::encode_string(value).into()
+            } else {
+                String::from_utf8_lossy(value)
```

Comment:
On the way into S2 when accepting input that is expected to be valid UTF-8, should fail if it isn't. On the way out from S2 for reads, ok to use `from_utf8_lossy`

Reply:
Into S2, it's pretty safe. The data-type is `String` so will be valid UTF-8.